### PR TITLE
Add script to update the openmmc and bootloader to v1.6.0

### PR DESCRIPTION
**Pull request by gustavosr8**

### CODE REVIEW

File: 08-mmc-upload/mmc-downgrade.sh
Comment (augustofg) [inline 08-mmc-upload/mmc-downgrade.sh:1]: Why is this necessary? Can't we have a single 'upload' script that you specify the MMC version? Sure, you will need to adjust the URL format for versions < 1.6.0, but there would be less duplicated code.

File: 08-mmc-upload/ipmi-overcharge.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+MCH_IP="$1"
+SLOT="$2"
+#BINARIE_OLD="$3"
+#BINARIE_NEW="$3"
+
+while [ 1 ]; do
+  #echo y | ipmitool -I lan -H ${MCH_IP} -A  none -T 0x82 -m 0x20 -t $((112 + 2*${SLOT})) hpm upgrade ${BINARIE_OLD} activate && break
+  #sleep 3
+  #echo y | ipmitool -I lan -H ${MCH_IP} -A  none -T 0x82 -m 0x20 -t $((112 + 2*${SLOT})) hpm upgrade ${BINARIE_NEW} activate && break
+  #sleep 3
```
Comment (augustofg) [inline 08-mmc-upload/ipmi-overcharge.sh]: Why these are commented? It would better if they are executed conditionally if the binaries are provided.

File: 08-mmc-upload/mmc-upload.sh [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+SCRIPTPATH="$( cd "$( dirname "${BASH_SOURCE[0]}"  )" && pwd  )"
+
+# Source FPGA mapping
+. ${SCRIPTPATH}/../misc/crate-fpga-mapping.sh
+
+MMC_BINARIES_DIR_REL=binaries
+MMC_BINARIES_DIR=${SCRIPTPATH}/${MMC_BINARIES_DIR_REL}
+
+OPENMMC_VERSION="v1.6.0"
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh]: I think it would be better to not hard-code the version here, but receive this as an argument, and deal with the idiosyncrasies of the URL for version 1.5.0.

File: 08-mmc-upload/mmc-upload.sh [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    ;;
+    *)
+      exec_cmd "ERR  " echo "Invalid bitstream type: ${bpm_crate_fpga}" >&2
+      exit 1
+    ;;
+  esac
+
+  tries=0
+
+  echo y | ~/sw/ipmitool/bin/ipmitool -I lan -H ${MCH_IP} -A  none -T 0x82 -m 0x20 -t $((112 + 2*${slot})) hpm upgrade ${binary} activate
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh]: This assumes a particular path for ipmitool. Change it to use the version in $PATH (aka don't hard-code the ipmitool path). If you want to use a different ipmitool binary, the user is responsible in inserting the path corresponding path at the beginning of the PATH variable before calling this script.

File: 08-mmc-upload/mmc-upload.sh [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      exit 1
+    ;;
+  esac
+  ipmitool -I lan -H ${MCH_IP} -A  none -T 0x82 -m 0x20 -t $((112 + 2*${slot})) raw 0x32 0x03 ${clk_cfg}
+  if [[ $? != 0 ]]; then
+    echo "Clock configuration failed! Slot ${slot} Crate ${MCH_IP}"
+    exit 1
+  fi
+  echo "Clock Switch configurated! Slot ${slot}"
+done
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh]: This should be done in a separate script, it is not necessary to run on every update.

File: 08-mmc-upload/mmc-upload.sh [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+#for slot in `seq 5 16`; do
+#  ./fru_soft_rst.expect ${MCH_IP} ${slot}
+#done
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh]: I think this is not necessary anymore. Even if it was a hard-reset, it is only required when changing the AFC FRU.

File: 08-mmc-upload/set-fru-info.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+i,c = np.where(component_list == args.rack_number)
+fru = Fru()
+fru.load_bin(args.bin_in)
+yaml = fru.to_dict()
+yaml['BoardInfo']['manufacturer']='Creotech'
+yaml['BoardInfo']['product_name']='AFC:4.0.2'
+yaml['BoardInfo']['serial_number']=component_list['SerialNumber'][i[0]]
+yaml['BoardInfo']['part_number']='AFC'
+yaml['BoardInfo']['fru_file_id']='AFCFRU'
+yaml['BoardInfo']['manufacturer']='CNPEM'
```
Comment (augustofg) [inline 08-mmc-upload/set-fru-info.py]: You are overwriting the previous (correct) value for BoardInfo.manufacturer (Creotech).

File: 08-mmc-upload/update-fru.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+ipmitool -I lan -H ${MCH_IP} -A none -T 0x82 -m 0x20 -t 116 fru read 0 ${FRU_BINARIES_DIR}/rack-${CRATE_NUMBER}-read.bin
+
+if [[ $? != 0 ]]; then
+	echo "FRU read failed. Crate ${MCH_IP}"
+	exit 1
+fi
+
+
+#. ${SCRIPTPATH}/set-fru-info.py --csv-in ${CSV_PATH} --rack_number ${CRATE_NUMBER} --bin-in ${FRU_BINARIES_DIR}/rack-${CRATE_NUMBER}-read.bin --bin-out ${FRU_BINARIES_DIR}/rack-${CRATE_NUMBER}-write.bin
```
Comment (augustofg) [inline 08-mmc-upload/update-fru.sh]: If this is not needed anymore, just remove it.

**Review by augustofg — CHANGES_REQUESTED** [review]

File: misc/crate-fpga-mapping.sh
Comment (augustofg) [inline misc/crate-fpga-mapping.sh:1]: For crates 23, 24 and 25 (development crates) I would keep the board mapping body 'empty' as these change frequently.

File: 08-mmc-upload/mmc-download.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+SCRIPTPATH="$( cd "$( dirname "${BASH_SOURCE[0]}"  )" && pwd  )"
+
+# Source FPGA mapping
+. ${SCRIPTPATH}/../misc/crate-fpga-mapping.sh
+
+MMC_BINARIES_DIR_REL=binaries
+MMC_BINARIES_DIR=${SCRIPTPATH}/${MMC_BINARIES_DIR_REL}
+
+OPENMMC_URL_RELEASE_BASE="https://github.com/lnls-dig/openMMC/releases/download"
+OPENMMC_VERSION="v1.6.0"
```
Comment (augustofg) [inline 08-mmc-upload/mmc-download.sh]: openMMC version hard-coded here. I would make this script accept only version v1.6.0 and upwards (there is a simple way to do version string comparison in shell script?). Older versions are incompatible.

File: 08-mmc-upload/mmc-upload.sh [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    # Retry the update if the hash compare fails
+    tries=0
+    while [[ ${COMMIT_HASH_READ} != ${COMMIT_HASH} ]]
+    do
+        echo "Hash compare failed! Slot ${slot} Crate ${crate_ip}"
+        # Only trie 3 times
+        if [ $counter -gt 3 ]; then
+            echo "Read Hash: ${COMMIT_HASH_READ}"
+            echo "Gave up..."
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh]: Nit: "Giving up..." is more idiomatic.

File: misc/crate-fpga-mapping.sh
Comment (augustofg) [reply misc/crate-fpga-mapping.sh:1]: Also, aren't the array declarations for de-23 and de-24 lacking?

**Review by augustofg — CHANGES_REQUESTED** [review]

File: misc/crate-fpga-mapping.sh
Comment (gustavosr8) [reply misc/crate-fpga-mapping.sh:1]: Hmm, they are. I wonder if there is any reason for this... anyway, I will add their declarations and keep the mapping empty

File: 08-mmc-upload/mmc-download.sh (hunk 8 above)
Comment (gustavosr8) [reply 08-mmc-upload/mmc-download.sh]: Do you think that the version check should only be on this script or every script that receives the version value must check on this?

File: 08-mmc-upload/mmc-download.sh [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+# Simple argument checking
+if [ $# -lt 1 ]; then
+    echo "Usage: ./mmc-download.sh <OpenMMC Version>"
+    exit 1
+fi
+OPENMMC_VERSION="$1"
+VERSION_NUM="${OPENMMC_VERSION#v}"
+
+if [[ "$(printf '%s\n' $VERSION_NUM 1.6.0 | sort -V | head -n 1)" != 1.6.0 ]]; then
```
Comment (augustofg) [inline 08-mmc-upload/mmc-download.sh:21]: Clever.

File: 08-mmc-upload/update-fru.sh [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+. ${SCRIPTPATH}/../misc/crate-fpga-mapping.sh
+
+FRU_BINARIES_DIR_REL=fru_binaries
+FRU_BINARIES_DIR=${SCRIPTPATH}/${FRU_BINARIES_DIR_REL}
+mkdir -p ${FRU_BINARIES_DIR}
+
+set +u
+
+# Simple argument checking
+if [ $# -lt 2 ]; then
```
Comment (augustofg) [inline 08-mmc-upload/update-fru.sh]: Shouldn't it be `if [ $# -lt 1 ]; then` ?

File: 08-mmc-upload/config-clkswitch.sh [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+# Source FPGA mapping
+. ${SCRIPTPATH}/../misc/crate-fpga-mapping.sh
+
+# Source common functions
+. ${SCRIPTPATH}/../misc/functions.sh
+
+set +u
+
+# Simple argument checking
+if [ $# -lt 2 ]; then
```
Comment (augustofg) [inline 08-mmc-upload/config-clkswitch.sh]: `if [ $# -lt 1 ]; then` maybe ?

File: 08-mmc-upload/mmc-upload.sh [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+set -u
+
+# Expect the firmwares to be in the binaries diretory
+MMC_BINARIES_DIR_REL=binaries
+MMC_BINARIES_DIR=${SCRIPTPATH}/${MMC_BINARIES_DIR_REL}
+AFCv3_BINARY="openMMC-afcv3.1-8sfp-${OPENMMC_VERSION}.hpm"
+AFCv4_BINARY="openMMC-afcv4.0-lamp-${OPENMMC_VERSION}.hpm"
+
+# Read the Commit SHA from the given firmware version
+COMMIT_HASH=$(git ls-remote https://github.com/lnls-dig/openMMC.git refs/tags/${OPENMMC_VERSION} | cut -f1 | tr -d "\n")
```
Comment (augustofg) [inline 08-mmc-upload/mmc-upload.sh:33]: noice.

**Review by augustofg — CHANGES_REQUESTED** [review]
Nit: Replace use consistent indentation, chose if you want to use spaces or tabs. I see both are used in some files.

File: 08-mmc-upload/mmc-download.sh (hunk 8 above)
Comment (augustofg) [reply 08-mmc-upload/mmc-download.sh]: I think just here is enough.

**Review by augustofg — APPROVED** [review]
